Sound mode support media_player general and denonavr

starkillerOG opened the pull request:
## Description:
Added backend support for selecting the sound mode and updating the sound mode status.
This is general support for all media_player components.
This box is not shown in media_players that did not (yet) implement sound mode support in the backend.
I implemented the sound mode support (backend) for the denonavr platform and that one works perfictly.
I already have a pull request for the frontend to have a dropdown select box in the user interface if the media_player supports sound mode.
home-assistant/home-assistant-polymer#815

I tested this code and it works on Hassbian on my pi.

On the home assistant chat I heard that other platforms like onkyo and sonos could also implement sound mode support and probably most receivers could support this, this also provides a base for other platforms to simply implement sound mode support.

Pull request in home-assistant.github.io with documentation (if applicable): home-assistant/home-assistant.github.io#4470


## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  - platform: denonavr
    host: IP_ADRESS
    name: "Marantz"
    show_all_sources: False
    timeout: 2
    sound_mode: True
    sound_mode_dict: {'MUSIC':['PLII MUSIC'], 'MOVIE':['PLII MOVIE'], 'GAME':['PLII GAME'], 'PURE DIRECT':['DIRECT'], 'AUTO':['None'], 'DOLBY DIGITAL':['DOLBY DIGITAL'], 'MCH STEREO':['MULTI CH STEREO'], 'STEREO':['STEREO']}
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

starkillerOG committed:
Add sound mode support

add sound mode support using new denonavr library functions

starkillerOG committed:
Denonavr to v.0.7.0

starkillerOG committed:
Add sound mode support

starkillerOG committed:
Add sound mode support

starkillerOG committed:
Add sound mode support

starkillerOG commented:
Can someone please merge this PR?
It has already been discussed in [this old PR](https://github.com/home-assistant/home-assistant/pull/11764) .
It is now using the new denonavr v0.7.0 so all new code is in that library as was required by the revieuwers.

starkillerOG commented:
@rytilahti could you review this new code which uses the denonavr library v.0.7.0?

starkillerOG commented:
@scarface-4711 could you maybe review this PR, I would really like to wrap this up.

starkillerOG commented:
@balloob Could you please take a look at this PR (since you merged the last denonavr PR)?
Waiting for a revieuw for a month now, really would like to finish this PR.
I am really confedent that it can be merged fast, since previous versions have already been looked at and revieuwd.

scarface-4711 commented on homeassistant/components/media_player/denonavr.py:
```diff
@@ -197,12 +229,27 @@ def source_list(self):
+        return self._sound_mode
+
+    @property
+    def sound_mode_raw(self):
```
Same as for **self._sound_mode_raw**. There is no equivalent property in constructor of parent media_player class, thus I assume it is not used.

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -106,6 +109,7 @@
 SUPPORT_CLEAR_PLAYLIST = 8192
 SUPPORT_PLAY = 16384
 SUPPORT_SHUFFLE_SET = 32768
+SUPPORT_SELECT_SOUND_MODE = 65536
```
We should not add this as a new supported feature for media players. We should only add things that exist among many media players.

Revert all changes to this file and add a new service to the denonavr platform called `media_player.denonavr_set_sound_mode` to allow changing thet mode.

rytilahti replied:
I was planning to add support for sound modes in a form or another to songpal, so that'd be two platforms for using such a feature. I haven't checked out yet how to do that, but will try to find some time during the upcoming weekend for it, if that's fine?

starkillerOG replied:
@balloob I asked about this on the form when I first started with this code, and at that time there was also someone who was intrested in implementing sound mode support for the sonos platform based on my code (but he wanted to wait until I finished and it has been quite a long time).

But that would already make 3 platforms that would uses this.
Besides almost all sound systems have some kind of sound mode feature, therefore it could be implemented in most of the platforms if someone is willing to make the code (and has that device).

Therfore I made this a general support feature in the media_player/__init__.py file, and I think this is a good thing. Do you agree @balloob?

balloob replied:
See https://developers.home-assistant.io/docs/en/entity_index.html#changing-the-entity-model

balloob replied:
So let me close this PR and the frontend PR, and you can open an issue on the architecture repo to propose the new feature.

starkillerOG committed:
Added sound_mode_raw as a state attribute

This is usefull because the sound_mode_raw shows more information than only the sound_mode.

starkillerOG committed:
update sound mode raw property

starkillerOG committed:
spacing

starkillerOG committed:
spacing

starkillerOG commented:
Tested the new commit (sound mode raw as state attribute), and it all works.
@scarface-4711, @balloob, anything else I schould change?
@balloob do you agree with keeping the general sound mode support for the other platforms (in the media_player/__init__.py file?

starkillerOG commented:
@balloob How long is it going to take to get this merged if I go and open an issue on the architecture repo?
In fact everything in the general files (media_player/__init__.py, media_player/services.yaml) is just an coppy of the already implemented `ATTR_INPUT_SOURCE = 'source'` and `ATTR_INPUT_SOURCE_LIST = 'source_list'` attributes. Therefore the architecture is identical to this already implemented feature, so I do not see much problems.

I have been working on this sound mode support [since january](https://github.com/home-assistant/home-assistant/pull/11764), and it was already working in january. I just experiance so much resistance to get this implemented. I am verry willing to implement changes, no problem, I just need some more detailed directions of what is expected.

So if this is going to take again so long to get it through the architecturee repo, I am not going to do that. Maybe I could then converted it to a denonavr specific service, and otherwise I will just continue as I have it now and use it as a custom component. I would really like to help the home assistant community implement some new features (such as this), but if you get so much resistance the fun fades away fast.

rytilahti commented:
Is this PR now active again, that is, should I base songpal modifications on top of this PR?

starkillerOG commented:
I will probably make some small adjustments to this PR based on the help of @balloob.
But I think we are continuing with this PR.

rytilahti commented:
Ok, could you please create a separate (rebased) PR for the sound mode support in general, and another one for denonavr support? Then I could base my WIP on that generic support PR.

balloob commented:
Yes go for it.
